### PR TITLE
feat: create web view of ware house.

### PR DESCRIPTION
**Pull request by nmimsnikhil**
[Asana](https://app.asana.com/0/1192407899754728/1198220653997629)
step2
step4
step5
warehouse1
warehouse2

### CODE REVIEW

File: erpnext/stock/doctype/warehouse/warehouse.py [hunk 1]
```diff
@@ -32,6 +32,36 @@ def onload(self):
 
 	def on_update(self):
 		self.update_nsm_model()
+		self.create_warehouse_address()
+		self.create_warehouse_contact()
+
+	def create_warehouse_address(self):
```
Comment (rvpasari) [inline erpnext/stock/doctype/warehouse/warehouse.py:39]: docstrings and comments please. Every function should have a docstring and every block inside it should have a header comment of sorts

File: erpnext/stock/doctype/warehouse/warehouse.py [hunk 2]
```diff
@@ -32,6 +32,36 @@ def onload(self):
...
+				"link_name": self.customer
+			}, fields=["parent as name"])
+
+			for address_name in address_names:
+				address = frappe.get_doc('Address', address_name.get('name'))
+				if not address.has_link('Warehouse', self.name):
+					address.append('links', dict(link_doctype='Warehouse', link_name=self.name))
+					address.save()
+
+	def create_warehouse_contact(self):
```
Comment (rvpasari) [inline erpnext/stock/doctype/warehouse/warehouse.py:53]: Same here

File: erpnext/stock/doctype/warehouse/warehouse.py [hunk 3]
```diff
@@ -199,4 +229,39 @@ def get_warehouses_based_on_account(account, company=None):
 		frappe.throw(_("Warehouse not found against the account {0}")
 			.format(account))
 
-	return warehouses
\ No newline at end of file
+	return warehouses
+
+def get_warehouse_list(doctype, txt, filters, limit_start, limit_page_length=20, order_by="modified"):
```
Comment (rvpasari) [inline erpnext/stock/doctype/warehouse/warehouse.py]: This requires an elaborate docstring explaining the arguments and returned variables. Please refer to the docstring extension for vscode pinned to the gurukul page, and implement all docstrings leaving the first line (the one containing the opening quotations) blank

**Review by rvpasari — CHANGES_REQUESTED** [review]

**Comment by hrwX** [comment]
@nmimsnikhil do not make any custom development, use the list.html for this purpose

**Comment by nmimsnikhil** [comment]
@sahil28297 and @hrwX  please ignore docstring.

File: erpnext/stock/doctype/warehouse/warehouse.py [hunk 4]
```diff
@@ -199,4 +229,48 @@ def get_warehouses_based_on_account(account, company=None):
...
+		return docs
+
+	elif customer :
+		docs = frappe.db.sql('''SELECT * FROM `tabBin`
+		WHERE warehouse=(SELECT name FROM `tabWarehouse`
+		WHERE customer=%s AND published=1)''', (customer), as_dict = True)
+		for doc in docs:
+			item_name = frappe.get_value("Item", doc.item_code, "item_name")
+			doc = doc.update({"item_name": item_name})
+		return docs
```
Comment (hrwX) [inline erpnext/stock/doctype/warehouse/warehouse.py]: - Commonify this.
- There shouldn't be an explicit check for Admin, if there is a customer with admin user, only then it should be visible

**Review by hrwX — CHANGES_REQUESTED** [review]

File: erpnext/stock/doctype/warehouse/warehouse.py [hunk 5]
```diff
@@ -199,4 +229,42 @@ def get_warehouses_based_on_account(account, company=None):
...
-	return warehouses
\ No newline at end of file
+	return warehouses
+
+def get_warehouse_list(doctype, txt, filters, limit_start, limit_page_length=20, order_by="modified"):
+	"""Getting list of warehouse and there items details from Warehouse and Bin doctype."""
+	user = frappe.session.user
+	customer = frappe.get_value("User", user, "full_name")
+	flt = {"published": 1}
+	if customer != "Administrator":
```
Comment (hrwX) [inline erpnext/stock/doctype/warehouse/warehouse.py]: - Technically you need to get the customer to which the user `test@user.com` is linked to in dynamic links.
- The current implementation is not correct, do one thing, from the session user, check the customer it has access to, and then get the warehouses of that customer

**Review by hrwX — CHANGES_REQUESTED** [review]

File: erpnext/stock/doctype/warehouse/warehouse.py [hunk 6]
```diff
@@ -199,4 +229,54 @@ def get_warehouses_based_on_account(account, company=None):
...
\ No newline at end of file
+	return warehouses
+
+def get_warehouse_list(doctype, txt, filters, limit_start, limit_page_length=20, order_by="modified"):
+	"""Getting list of warehouse and there items details from Warehouse and Bin doctype."""
+	user = frappe.session.user
+	customers = []
+	docs = []
+	# Checking user have 'Warehouse' permission or not.
+	if frappe.has_permission("Warehouse"):
```
Comment (hrwX) [inline erpnext/stock/doctype/warehouse/warehouse.py]: Lets remove this, just saw, `allow_guest_to_view` in doctype json

File: erpnext/stock/doctype/warehouse/warehouse.py [hunk 7]
```diff
@@ -199,4 +229,54 @@ def get_warehouses_based_on_account(account, company=None):
...
+		for doc in frappe.get_all("Contact", {"user": user}):
```
Comment (hrwX) [inline erpnext/stock/doctype/warehouse/warehouse.py]:
```suggestion
		for doc in frappe.get_all("Contact", {"email_id": frappe.session.user}):
```
Just check whether frappe.session.user is email

File: erpnext/stock/doctype/warehouse/warehouse.py [hunk 8]
```diff
@@ -199,4 +229,54 @@ def get_warehouses_based_on_account(account, company=None):
...
+			# getting list of customers tha have user assigned through contact. 
+			customers += frappe.get_all("Dynamic Link", {
+				"parent": doc.name,
+				"link_doctype": "Customer"
+			}, "link_name")
```
Comment (hrwX) [inline erpnext/stock/doctype/warehouse/warehouse.py]: directly use sql, and join Contact and Dynamic Link on name and parent

File: erpnext/stock/doctype/warehouse/warehouse.py [hunk 9]
```diff
@@ -199,4 +229,54 @@ def get_warehouses_based_on_account(account, company=None):
...
+
+		for customer in customers:
+			# Getting user warehouse.
+			flt = {
+				"published": 1,
+				"customer": customer.link_name
+			}
+			warehouses = frappe.get_all("Warehouse", flt)
```
Comment (hrwX) [inline erpnext/stock/doctype/warehouse/warehouse.py]: create a list of customer and do an `IN` query

File: erpnext/stock/doctype/warehouse/warehouse.py [hunk 10]
```diff
@@ -199,4 +229,54 @@ def get_warehouses_based_on_account(account, company=None):
...
+
+		for warehouse in warehouses:
+			# Fetching items details.
+			docs += frappe.get_all("Bin",{"warehouse": warehouse.name}, "*")
```
Comment (hrwX) [inline erpnext/stock/doctype/warehouse/warehouse.py]: create a list of warehouse and do an `IN` query

File: erpnext/stock/doctype/warehouse/warehouse.py [hunk 11]
```diff
@@ -199,4 +229,54 @@ def get_warehouses_based_on_account(account, company=None):
...
+
+		for doc in docs:
+			item_name = frappe.get_value("Item", doc.item_code, "item_name")
+			doc = doc.update({"item_name": item_name})
```
Comment (hrwX) [inline erpnext/stock/doctype/warehouse/warehouse.py]: What is this update doing ?? You aren't event saving the doc

**Review by hrwX — CHANGES_REQUESTED** [review]

File: erpnext/stock/doctype/warehouse/warehouse.py (hunk 11 above)
Comment (nmimsnikhil) [reply erpnext/stock/doctype/warehouse/warehouse.py]: its append the item_name to the returning docs

File: erpnext/stock/doctype/warehouse/warehouse.py (hunk 6 above)
Comment (nmimsnikhil) [reply erpnext/stock/doctype/warehouse/warehouse.py]: sorry allow_guest to view would be 0.
as only user with customer role allow to see their warehouse.

**Review by hrwX — APPROVED** [review]

**Review by sahil28297 — APPROVED** [review]

**Pull request merged**